File: backend/api/websocket.py
#!/usr/bin/env python3
"""
WebSocket event handlers
"""

from flask_socketio import emit, join_room, leave_room


def register_socket_events(socketio):
    """Register WebSocket event handlers"""
    
    @socketio.on('connect')
    def handle_connect():
        """Handle client connection"""
        print('Client connected')
        emit('connected', {'message': 'Connected to server'})
    
    @socketio.on('disconnect')
    def handle_disconnect():
        """Handle client disconnection"""
        print('Client disconnected')
    
    @socketio.on('join')
    def handle_join(data):
        """Join a room (for job-specific updates)"""
        room = data.get('room')
        if room:
            join_room(room)
            emit('joined', {'room': room}, room=room)
    
    @socketio.on('leave')
    def handle_leave(data):
        """Leave a room"""
        room = data.get('room')
        if room:
            leave_room(room)
            emit('left', {'room': room}, room=room)


# Helper functions for emitting updates
def emit_job_update(socketio, job_id, data):
    """Emit job update to specific room"""
    room = f'job_{job_id}'
    socketio.emit('job_update', data, room=room)


def emit_progress(socketio, job_id, progress, step):
    """Emit progress update"""
    room = f'job_{job_id}'
    socketio.emit('progress', {
        'job_id': job_id,
        'progress': progress,
        'step': step
    }, room=room)


def emit_log(socketio, job_id, message, level='info'):
    """Emit log message"""
    room = f'job_{job_id}'
    socketio.emit('log', {
        'job_id': job_id,
        'message': message,
        'level': level
    }, room=room)


def emit_error(socketio, job_id, error):
    """Emit error message"""
    room = f'job_{job_id}'
    socketio.emit('error', {
        'job_id': job_id,
        'error': error
    }, room=room)


def emit_complete(socketio, job_id, clips_created):
    """Emit job completion"""
    room = f'job_{job_id}'
    socketio.emit('complete', {
        'job_id': job_id,
        'clips_created': clips_created
    }, room=room)

File: frontend/src/services/websocket.ts
import { io, Socket } from "socket.io-client";

const WS_URL = process.env.NEXT_PUBLIC_WS_URL || "http://localhost:5001";

class WebSocketClient {
  private socket: Socket | null = null;
  private reconnectAttempts = 0;
  private maxReconnectAttempts = 5;

  connect(): Socket {
    if (this.socket?.connected) {
      return this.socket;
    }

    this.socket = io(WS_URL, {
      transports: ["websocket", "polling"],
      reconnectionAttempts: this.maxReconnectAttempts,
    });

    this.socket.on("connect", () => {
      console.log("WebSocket connected");
      this.reconnectAttempts = 0;
    });

    this.socket.on("disconnect", () => {
      console.log("WebSocket disconnected");
    });

    this.socket.on("connect_error", (error) => {
      console.error("WebSocket connection error:", error);
      this.reconnectAttempts++;
    });

    return this.socket;
  }

  disconnect(): void {
    if (this.socket) {
      this.socket.disconnect();
      this.socket = null;
    }
  }

  joinJobRoom(jobId: number): void {
    if (this.socket) {
      this.socket.emit("join", { room: `job_${jobId}` });
    }
  }

  leaveJobRoom(jobId: number): void {
    if (this.socket) {
      this.socket.emit("leave", { room: `job_${jobId}` });
    }
  }

  onJobUpdate(callback: (data: unknown) => void): void {
    if (this.socket) {
      this.socket.on("job_update", callback);
    }
  }

  onProgress(
    callback: (data: { job_id: number; progress: number; step: string }) => void
  ): void {
    if (this.socket) {
      this.socket.on("progress", callback);
    }
  }

  onLog(
    callback: (data: { job_id: number; message: string; level: string }) => void
  ): void {
    if (this.socket) {
      this.socket.on("log", callback);
    }
  }

  onError(callback: (data: { job_id: number; error: string }) => void): void {
    if (this.socket) {
      this.socket.on("error", callback);
    }
  }

  onComplete(
    callback: (data: { job_id: number; clips_created: number }) => void
  ): void {
    if (this.socket) {
      this.socket.on("complete", callback);
    }
  }

  removeAllListeners(): void {
    if (this.socket) {
      this.socket.removeAllListeners();
    }
  }
}

export const websocket = new WebSocketClient();
export default websocket;
